components/modules/pulsecnt.c

```diff
@@ -0,0 +1,732 @@
+  } else {
+    if (pc->is_debug) ESP_LOGI("pulsecnt", "Could not find cb for unit %d with ptr %d", unit, pc->cb_ref);
+  }
+}
+
+// Get the pulsecnt.pctr object from the stack which is the struct pulsecnt_t
+static pulsecnt_t pulsecnt_get( lua_State *L, int stack )
+{
+  return (pulsecnt_t)luaL_checkudata(L, stack, "pulsecnt.pctr");
+}
+
+// Lua: pc:setFilter(clkCyclesToIgnore)
+// Example: pc:setFilter(100) -- Ignore any signal shorter than 100 clock cycles. 80Mhz clock.
+// You can ignore from 0 to 1023 clock cycles
+static int pulsecnt_set_filter( lua_State *L ) {
+  int stack = 0;
+
+  // when we're called from an object the stack index 1 has our self ref 
+  pulsecnt_t pc = pulsecnt_get(L, ++stack);
+
+  // Get clkCyclesToIgnore -- first arg after self arg
+  int clkCyclesToIgnore = luaL_checkinteger(L, ++stack);
+  luaL_argcheck(L, clkCyclesToIgnore >= 0 && clkCyclesToIgnore <= 1023, stack, "The clkCyclesToIgnore number allows 0 to 1023");
+
+  pcnt_set_filter_value(pc->unit, clkCyclesToIgnore);
+
+  if (pc->is_debug) ESP_LOGI("pulsecnt", "Setup filter for unit %d with clkCyclesToIgnore %d", pc->unit, clkCyclesToIgnore);
+
+  return 0;
+
+}
+
+// Lua: pc:setThres(thresh0_val, thresh1_val)
+// Example: pc:setThres(-5, 5)
+// When you set the threshold, the pulse counter will be reset and the callback will be attached.
+static int pulsecnt_set_thres( lua_State *L ) {
+  int stack = 0;
+
+  // when we're called from an object the stack index 1 has our self ref 
+  pulsecnt_t pc = pulsecnt_get(L, ++stack);
+
+  // Get thresh0_val -- first arg after self arg
+  int thresh0_val = luaL_checkinteger(L, ++stack);
+  luaL_argcheck(L, thresh0_val >= -32768 && thresh0_val <= 32767, stack, "The thresh0_val number allows -32768 to 32767");
+
+  // Get thresh0_val -- first arg after self arg
+  int thresh1_val = luaL_checkinteger(L, ++stack);
+  luaL_argcheck(L, thresh1_val >= -32768 && thresh1_val <= 32767, stack, "The thresh1_val number allows -32768 to 32767");
+
+  // store it so we have it during callback
+  pc->thresh0 = thresh0_val;
+  pc->thresh1 = thresh1_val;
+
+  /* Set threshold 0 and 1 values and enable events to watch */
+  pcnt_set_event_value(pc->unit, PCNT_EVT_THRES_0, thresh0_val);
+  pcnt_event_enable(pc->unit, PCNT_EVT_THRES_0);
+  pcnt_set_event_value(pc->unit, PCNT_EVT_THRES_1, thresh1_val);
+  pcnt_event_enable(pc->unit, PCNT_EVT_THRES_1);
+  /* Enable events on zero, maximum and minimum limit values */
+  pcnt_event_enable(pc->unit, PCNT_EVT_ZERO);
+  pcnt_event_enable(pc->unit, PCNT_EVT_H_LIM);
+  pcnt_event_enable(pc->unit, PCNT_EVT_L_LIM);
+
+  /* Initialize PCNT's counter */
+  pcnt_counter_pause(pc->unit);
+  pcnt_counter_clear(pc->unit);
+
+  // check if there's a callback otherwise don't trigger interrupt, instead they may just be polling
+  if (pc->cb_ref != LUA_NOREF) {
+    /* Register ISR handler and enable interrupts for PCNT unit */
+    pcnt_isr_register(pulsecnt_intr_handler, NULL, 0, &user_isr_handle);
```

Comment:
Check the return code from `pcnt_isr_register()`.
It can be registered multiple times. You could register only if `user_isr_handle` is not `NULL`.
Also see below.